python/ray/dag/compiled_dag_node.py

```diff
@@ -150,54 +138,58 @@ def _wrap_exception(exc):
     return wrapped
 
 
-def _exec_task(self, task: "ExecutableTask", idx: int) -> bool:
+def _select_next_nodes(
+    actor_to_candidates: Dict["ray.actor.ActorHandle", List[DAGOperationGraphNode]],
+    graph: Dict[int, Dict[DAGNodeOperationType, DAGOperationGraphNode]],
+):
     """
-    Execute the task.
+    This function selects the next nodes for topological sort to generate execution
+    schedule. If there are multiple DAGOperationGraphNodes with zero in-degree,
+    select nodes based on the following rules:
+
+    #1  If the nodes are not NCCL write nodes, select the one with the smallest
+        `bind_index`. If there are multiple candidate nodes with the smallest
+        `bind_index` of the actors that they belong to, any one of them is
+        acceptable. For the implementation details, we maintain a priority queue
+        for each actor, where the head of the priority queue is the node with the
+        smallest `bind_index`.
+    #2  If #1 cannot be satisfied, it means that all candidate nodes are
+        NCCL write nodes. In this case, select the one that is the head of the
+        priority queue and its downstream nodes, regardless of whether the
```

Comment:
downstream nodes are always READ right? can you briefly mention this?

Reply:
Updated https://github.com/ray-project/ray/pull/46911/commits/50ff5c1385be55a1220fd8b9ef05f28b89444c03